Source/JavaScriptCore/heap/FreeList.cpp

```diff
@@ -39,53 +39,50 @@ FreeList::~FreeList()
 
 void FreeList::clear()
 {
-    m_scrambledHead = 0;
+    m_head = reinterpret_cast_ptr<FreeCell*>(1);
     m_secret = 0;
+    m_payloadStart = nullptr;
     m_payloadEnd = nullptr;
-    m_remaining = 0;
     m_originalSize = 0;
 }
 
-void FreeList::initializeList(FreeCell* head, uintptr_t secret, unsigned bytes)
+void FreeList::initialize(FreeCell* start, uint64_t secret, unsigned bytes)
 {
-    // It's *slightly* more optimal to use a scrambled head. It saves a register on the fast path.
-    m_scrambledHead = FreeCell::scramble(head, secret);
+    if (UNLIKELY(!start)) {
+        clear();
+        return;
+    }
+    auto [offsetToNext, lengthInBytes] = start->decode(secret);
+    m_head = bitwise_cast<FreeCell*>(bitwise_cast<char*>(start) + offsetToNext);
     m_secret = secret;
-    m_payloadEnd = nullptr;
-    m_remaining = 0;
+    m_payloadStart = reinterpret_cast<char*>(start);
+    m_payloadEnd = m_payloadStart + lengthInBytes;
     m_originalSize = bytes;
-}
-
-void FreeList::initializeBump(char* payloadEnd, unsigned remaining)
-{
-    m_scrambledHead = 0;
-    m_secret = 0;
-    m_payloadEnd = payloadEnd;
-    m_remaining = remaining;
-    m_originalSize = remaining;
+    RELEASE_ASSERT(m_payloadEnd - m_payloadStart < (ptrdiff_t)MarkedBlock::blockSize);
 }
 
 bool FreeList::contains(HeapCell* target) const
 {
-    if (m_remaining) {
-        const void* start = (m_payloadEnd - m_remaining);
-        const void* end = m_payloadEnd;
-        return (start <= target) && (target < end);
-    }
+    char* targetPtr = bitwise_cast<char*>(target);
+    if (m_payloadStart <= targetPtr && targetPtr < m_payloadEnd)
+        return true;
 
     FreeCell* candidate = head();
-    while (candidate) {
-        if (bitwise_cast<HeapCell*>(candidate) == target)
+    while (!isSentinel(candidate)) {
+        auto [offsetToNext, lengthInBytes] = candidate->decode(m_secret);
+        char* end = bitwise_cast<char*>(candidate + 1);
+        char* start = end - lengthInBytes;
```

Comment:
This is wrong.  `start` should be `bitwise_cast<char*>(candidate)`, and `end` should be `start + lengthInBytes`.  You're doing this computation in so many places that maybe you should have an ALWAYS_INLINE helper function to do them.

Reply:
Fixed the `contains()` method in latest push, but I still need to refactor this code duplication out.

Reply:
Refactored to use a single `FreeCell::advance()` function that loads and decodes the current interval, sets up the interval start and end, and computes the next next interval in a single call. Should resolve the code duplication issue.